src/control/PIDControl.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include "../Util.h"
+#include "../world_interface/data.h"
+
+#include <chrono>
+#include <functional>
+#include <optional>
+#include <utility>
+
+#include <Eigen/Core>
+
+namespace control {
+
+/**
+ * @brief A set of PID gains.
+ * @see https://en.wikipedia.org/wiki/PID_controller
+ */
+struct PIDGains {
+	double kP, kI, kD, kF;
+	std::optional<double> iZone;
+};
+
+/**
+ * @brief A PID controller for multiple independent dimensions.
+ *
+ * @tparam dim The number of dimensions to control.
+ * @see https://en.wikipedia.org/wiki/PID_controller
+ */
+template <int dim> class PIDController {
+public:
+	template <int d> using Arrayd = Eigen::Array<double, d, 1>;
+
+	/**
+	 * @brief Construct a new PIDController.
+	 *
+	 * @param gains The gains to use for the controller.
+	 */
+	PIDController(const std::array<PIDGains, dim>& gains) {
+		reset();
+		setGains(gains);
+	}
+
+	/**
+	 * @brief Set the gains of the PID controller.
+	 *
+	 * @param gains The new gains to set.
+	 */
+	void setGains(const std::array<PIDGains, dim>& gains) {
+		for (int i = 0; i < dim; i++) {
+			kP(i) = gains[i].kP;
+			kI(i) = gains[i].kI;
+			kD(i) = gains[i].kD;
+			kF(i) = gains[i].kF;
+			iZone(i) = gains[i].iZone.value_or(std::numeric_limits<double>::infinity());
+		}
+	}
+
+	/**
+	 * @brief Check if a target is currently set.
+	 *
+	 * @return bool True iff the controller currently has a target.
+	 */
+	bool hasTarget() {
+		return setPoint.has_value();
+	}
+
+	/**
+	 * @brief Set the target of the controller.
+	 *
+	 * @param setPoint An array of setpoints, one for each dimension.
+	 */
+	void setTarget(const Arrayd<dim>& setPoint) {
+		this->setPoint = setPoint;
+	}
+
+	/**
+	 * @brief Get the output from the controller.
```

Comment:
Instead of saying output from the controller, you can say "Get the dimension targets from the controller" so it's less repetitive with the method name.